src/mon/PGMap.cc

```diff
@@ -2390,12 +2390,16 @@ void PGMap::get_health_checks(
 
   // TOO_MANY_PGS
   if (num_in && cct->_conf->mon_pg_warn_max_per_osd > 0) {
-    int per = sum_pg_up / num_in;
-    if (per > cct->_conf->mon_pg_warn_max_per_osd) {
-      ostringstream ss;
-      ss << "too many PGs per OSD (" << per
-	 << " > max " << cct->_conf->mon_pg_warn_max_per_osd << ")";
-      checks->add("TOO_MANY_PGS", HEALTH_WARN, ss.str());
+    for (auto osd_num_pg : num_pg_by_osd) {
+      auto osd = osd_num_pg.first;
+      auto num_pg = osd_num_pg.second.acting;
+      if (num_pg > cct->_conf->mon_pg_warn_max_per_osd) {
+	ostringstream ss;
+	ss << "too many PGs on OSD." << osd << " (" << num_pg
+	   << " > max " << cct->_conf->mon_pg_warn_max_per_osd << ")";
+	checks->add("TOO_MANY_PGS", HEALTH_WARN, ss.str());
```

Comment:
Using `add` will assert if the check already exists.

What you probably want to do here is build a list of problematic OSDs, and then create the health check one time with the `detail` field containing the specific OSD ids.

Reply:
yes, that's why i break at seeing the first osd breach the upper bound. is it really necessary to list all of the problematic OSDs? how about just list part of them, like the first 10 OSDs?

Reply:
we do that with pgs.. if you look in PGMap.cc you'll see we have a limit to the number of detail items (pgs)